src/docgen/schema.ts

```diff
@@ -0,0 +1,335 @@
+/**
+ * Describes a type.
+ */
+export interface TypeSchema {
+
+  /**
+   * The type FQN. If missing, the type is a reference (array, map..)
+   */
+  readonly fqn?: string;
+
+  /**
+   * The name of the type (Map, List...)
+   */
+  readonly name: string;
+
+  /**
+   * The various types of the reference.
+   */
+  readonly types?: TypeSchema[];
+}
+
+/**
+ * Describes a property.
+ */
+export interface PropertySchema {
+
+  /**
+   * Unique id of the property.
+   */
+  readonly id: string;
+
+  /**
+   * Name.
+   */
+  readonly name: string;
+
+  /**
+   * Whether or not the property is optional.
+   */
+  readonly optional: boolean;
+
+  /**
+   * Whether or not the property is deprecated.
+   */
+  readonly deprecated: boolean;
+
+  /**
+   * Deprecation reason (if applicable)
+   */
+  readonly deprecationReason?: string;
+
+  /**
+   * Doc strings of the property.
+   */
+  readonly docs: string;
+
+  /**
+   * The type of the property.
+   */
+  readonly type: TypeSchema;
+
+  /**
+   * The default value of the property
+   */
+  readonly default?: string;
+}
+
+/**
+ * Describes a parameter.
+ */
+export interface ParameterSchema extends PropertySchema {}
+
+/**
+ * Common properties of a method.
+ */
+export interface MethodSchemaBase {
+
+  /**
+   * Unique id of the method.
+   */
+  readonly id: string;
+
+  /**
+   * Code snippet to display.
+   */
+  readonly snippet: string;
+
+  /**
+   * Method parameters.
+   */
+  readonly parameters: ParameterSchema[];
+}
+
+/**
+ * Describes a constructor.
+ */
+export interface InitializerSchema extends MethodSchemaBase {}
+
+/**
+ * Describes a method.
+ */
+export interface MethodSchema extends MethodSchemaBase {
+
+  /**
+   * Method name.
+   */
+  readonly name: string;
+}
+
+/**
+ * Describes a class.
+ */
+export interface ClassSchema {
+
+  /**
+   * Unique class id.
+   */
+  readonly id: string;
+
+  /**
+   * Class name.
+   */
+  readonly name: string;
+
+  /**
+   * Interfaces this class implements.
+   */
+  readonly interfaces: TypeSchema[];
+
+  /**
+   * Doc string for the class.
+   */
+  readonly docs: string;
+
+  /**
+   * Class initializer.
+   */
+  readonly initializer?: InitializerSchema;
+
+  /**
+   * Instance methods.
+   */
+  readonly instanceMethods: MethodSchema[];
+
+  /**
+   * Static functions.
+   */
+  readonly staticFunctions: MethodSchema[];
+
+  /**
+   * Properties.
+   */
+  readonly properties: PropertySchema[];
+
+  /**
+   * Constants.
+   */
+  readonly constants: PropertySchema[];
+}
+
+/**
+ * Describes a construct.
+ */
+export interface ConstructSchema extends ClassSchema {}
+
+/**
+ * Describes a struct.
+ */
+export interface StructSchema {
+
+  /**
+   * Unique struct id.
+   */
+  readonly id: string;
+
+  /**
+   * Struct name.
```

Comment:
?

Reply:
`name` is the transliterated name of any given type, vs `id` or `fqn` which is obviously the same across languages.